### PR TITLE
fix bug in infer_base_unit. 

### PR DESCRIPTION
**Pull request by pmaunz**
infer_base_unit failed on quantities that were generated by multiplying the same dimension with different prefixes such as 'meter * millimeter'. This also leads to problems using .to_compact() as the base_unit of a quantity is inferred incorrectly
An example can be found in the added testcase.

### CODE REVIEW

File: pint/testsuite/test_numpy.py [hunk 1]
```diff
@@ -240,7 +240,7 @@ def test_reversible_op(self):
 
     def test_pickle(self):
         import pickle
-
+        set_application_registry(self.ureg)
```
Comment (hgrecco) [inline pint/testsuite/test_numpy.py]: I think this change belongs to a different PR

File: pint/testsuite/test_numpy.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import copy
 import operator as op
 
-from pint import DimensionalityError
+from pint import DimensionalityError, set_application_registry
```
Comment (hgrecco) [inline pint/testsuite/test_numpy.py]: I think this change belongs to a different PR

**Comment by hgrecco** [comment]
Thanks a lot. Sounds good to me! Could you please rebase to the develop branch? I am combining and testing together many patches before merging them to the master branch.

**Comment by pmaunz** [comment]
Thanks for the feedback. I closed this pull request and submitted two pull requests instead. 
`fix-pickle-unittest`  and
`fix-infer_base_unit` (where I have changed the code to use your udict instead of the defaultdict)

**Comment by hgrecco** [comment]
That was fast!
